gptqmodel/models/moe_lifecycle.py

```diff
@@ -0,0 +1,530 @@
+            model_class: The model class (to access module_tree)
+        
+        Returns:
+            The MoE block module, or None if not found
+        
+        Example:
+            For GLM-4, module_tree has "mlp:moe", so this returns layer_module.mlp
+            For MiniMax-M2, module_tree has "block_sparse_moe:moe", so this returns layer_module.block_sparse_moe
+        """
+        # Get MoE module name from model definition
+        moe_module_name = model_class.get_moe_module_name()
+        
+        if moe_module_name is None:
+            log.error(f"No :moe flag found in module_tree for {model_class.__name__}")
+            return None
+        
+        # Get the module by name
+        moe_block = getattr(layer_module, moe_module_name, None)
+        
+        return moe_block
+    
+    def get_experts_module(self, moe_block: nn.Module, model_class: type) -> Optional[nn.Module]:
+        """
+        Extract experts module from MoE block by checking common attribute names.
+        
+        Args:
+            moe_block: The MoE block module
+            model_class: The model class (for compatibility, unused)
+        
+        Returns:
+            The experts module, or None if not found
+        
+        Example:
+            Returns moe_block.experts if it exists
+        """
+        # Use the helper to get the attribute name
+        name = self.get_experts_module_name(moe_block)
+        if name:
+            experts_module = getattr(moe_block, name)
+            return experts_module
+        
+        return None
+    
+    def get_experts_module_name(self, moe_block: nn.Module) -> Optional[str]:
+        """
+        Get the attribute name for experts module if it exists.
+        
+        Args:
+            moe_block: The MoE block module
+        
+        Returns:
+            The attribute name ('experts', 'expert_list', 'expert_modules', etc.) or None
+        
+        Example:
+            name = hooks.get_experts_module_name(moe_block)
+            if name:
+                experts = getattr(moe_block, name)
+        """
+        # Try common expert container attribute names
+        for name in ['experts', 'expert_list', 'expert_modules']:
+            if hasattr(moe_block, name):
+                return name
+        return None
+    
+    def get_shared_experts_module_name(self, moe_block: nn.Module) -> Optional[str]:
+        """
+        Get the attribute name for shared experts module if it exists.
+        
+        Args:
+            moe_block: The MoE block module
+        
+        Returns:
+            The attribute name ('shared_experts', 'shared_expert', etc.) or None
+        
+        Example:
+            name = hooks.get_shared_experts_module_name(moe_block)
+            if name:
+                shared_experts = getattr(moe_block, name)
+        """
+        # Try both singular and plural naming
+        for name in ['shared_experts', 'shared_expert']:
```

Comment:
Move `['shared_experts', 'shared_expert']` list var to `MoELifecycleHooks.shared_expert_block_names` so models can override this using custom MoELifecycleHooks in model def.